pkg/controller/container-runtime-config/helpers.go

```diff
@@ -496,15 +496,38 @@ func getValidBlockedRegistries(releaseImage string, imgSpec *apicfgv1.ImageSpec)
 	for i, reg := range imgSpec.RegistrySources.BlockedRegistries {
 		// if there is a match, return all the blocked registries except the one that matched and return an error as well
 		if reg == payloadReg {
-			blockedRegs = append(blockedRegs, imgSpec.RegistrySources.BlockedRegistries[i+1:]...)
-			return blockedRegs, fmt.Errorf("error adding %q to blocked registries, cannot block the registry being used by the payload", payloadReg)
+			// If the payload registry doesn't have mirror rules configured for it, then don't add it to the blocked registries list
+			// TODO maybe? Add an opt-in configuration to ICSP and check if that is set as well
+			if !payloadRegHasMirrorConfig(payloadReg, icspRules) {
+				blockedRegs = append(blockedRegs, imgSpec.RegistrySources.BlockedRegistries[i+1:]...)
+				return blockedRegs, fmt.Errorf("error adding %q to blocked registries, cannot block the registry being used by the payload", payloadReg)
+			}
+			// Log a warning that we are adding the payload registries to the blocked registries list as there are mirror rules for it
+			glog.Warningf("%q is the payload registry, but will add it to the list of blocked registries as there are mirror rules configured for it", reg)
+
 		}
 		// Was not a match to the registry being used by the payload, so add to valid blocked registries
 		blockedRegs = append(blockedRegs, reg)
 	}
 	return blockedRegs, nil
 }
 
+// payloadRegHasMirrorConfig returns true if the payload registry has mirror rules configured for it
+func payloadRegHasMirrorConfig(payloadReg string, icspRules []*apioperatorsv1alpha1.ImageContentSourcePolicy) bool {
+	if len(icspRules) == 0 || icspRules == nil {
+		return false
+	}
+	// Go through the mirror rules configured and check if a source matches the payload registry
+	for _, rule := range icspRules {
+		for _, mirror := range rule.Spec.RepositoryDigestMirrors {
+			if strings.Contains(mirror.Source, payloadReg) {
```

Comment:
(`Contains` is very imprecise. I get that this is a best-effort code to prevent breaking clusters, it just offends my aesthetic sensibilities :) .  https://github.com/openshift/runtime-utils/blob/8b8348d80d1d1e7b6cf06fb009d5965e0b55baa2/pkg/registries/registries.go#L13 can probably be used for an exact check.)

Reply:
fixed

Reply:
Somehow GitHub is not showing the code this comments on — but if this was was a “`reg` is blocking `payloadReg`”, I now see just a `==` compare; shouldn’t that be `ScopeIsNestedInsideScope(payloadReg, reg)`?

Reply:
… actually, `payloadReg` is the whole registry. But see the other comment, we ideally want to match any `reg` that affects `releaseRepo`.